File: python/Backtracking/word_search.py
from typing import List

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        """
        Approach:
        - Use backtracking (DFS) to search for the word in the board.
        - For each cell, start a DFS if the first letter matches.
        - At each step, mark the cell as visited (using a set), and try all four directions.
        - If the word is found, return True. If not, backtrack and try other paths.
        - Restore the cell's state after backtracking.

        Time Complexity: O(m * n * 4^L), where m and n are the board dimensions and L is the length of the word.
        Space Complexity: O(L), for the recursion stack and path set.
        """
        ROWS, COLS = len(board), len(board[0])
        path = set()

        def dfs(r, c, i):
            if i == len(word):
                return True
            if (r < 0 or c < 0 or r >= ROWS or c >= COLS or word[i] != board[r][c] or (r,c) in path):
                return False

            path.add((r, c))
            res = (dfs(r + 1, c, i + 1) or dfs(r - 1, c, i + 1) or dfs(r, c + 1, i + 1) or dfs(r, c - 1, i + 1)) 

            path.remove((r,c))
            return res
        
        for r in range(ROWS):
            for c in range(COLS):
                if dfs(r, c, 0):
                    return True
        return False

# -------------------
# Test Cases
# -------------------
def test_word_search():
    sol = Solution()
    # Test case 1: Example from Leetcode
    board1 = [
        ["A","B","C","E"],
        ["S","F","C","S"],
        ["A","D","E","E"]
    ]
    assert sol.exist(board1, "ABCCED") == True
    assert sol.exist(board1, "SEE") == True
    assert sol.exist(board1, "ABCB") == False
    # Test case 2: Single letter
    board2 = [["A"]]
    assert sol.exist(board2, "A") == True
    assert sol.exist(board2, "B") == False
    # Test case 3: Word longer than board
    board3 = [["A","B"],["C","D"]]
    assert sol.exist(board3, "ABCDE") == False
    print("All test cases passed!")

if __name__ == "__main__":
    test_word_search()

File: C++/Backtracking/word_search.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Approach:
 * - Perform DFS starting from every cell in the board.
 * - At each cell, check if the current letter matches the word[i].
 * - Mark the cell as visited and explore in all 4 directions.
 * - Backtrack by unmarking the cell if the path does not lead to a solution.
 *
 * Time Complexity: O(m * n * 4^L), where m = rows, n = cols, L = word.length()
 * Space Complexity: O(L), for recursion stack and path tracking.
 */

class Solution {
public:
    int ROWS, COLS;

    bool exist(vector<vector<char>>& board, string word) {
        ROWS = board.size();
        COLS = board[0].size();

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (dfs(board, word, r, c, 0)) {
                    return true;
                }
            }
        }
        return false;
    }

private:
    bool dfs(vector<vector<char>>& board, string& word, int r, int c, int i) {
        if (i == word.size()) return true;  // word is completely matched

        // If out of bounds or mismatch or already visited
        if (r < 0 || c < 0 || r >= ROWS || c >= COLS || board[r][c] != word[i]) {
            return false;
        }

        char temp = board[r][c]; // save the original character
        board[r][c] = '#';       // mark as visited

        // Explore in all 4 directions
        bool found =
            dfs(board, word, r + 1, c, i + 1) ||
            dfs(board, word, r - 1, c, i + 1) ||
            dfs(board, word, r, c + 1, i + 1) ||
            dfs(board, word, r, c - 1, i + 1);

        board[r][c] = temp;      // backtrack: restore the original character
        return found;
    }
};

/*
------------------------------------------------------
📄 Line-by-Line C++ Syntax Explanation:
------------------------------------------------------

#include <bits/stdc++.h>
    → Includes all standard C++ libraries (vector, string, iostream, etc.) in one line.

using namespace std;
    → So we don’t need to prefix std:: before every type.

class Solution { ... };
    → Defines a class named Solution.

int ROWS, COLS;
    → Member variables to store board dimensions.

bool exist(vector<vector<char>>& board, string word)
    → Public method that checks if the word exists in the board.

board.size()
    → Number of rows in the board.

board[0].size()
    → Number of columns in the board.

for (int r = 0; r < ROWS; r++) { ... }
    → Iterate through each row.

for (int c = 0; c < COLS; c++) { ... }
    → Iterate through each column.

if (dfs(board, word, r, c, 0)) { return true; }
    → Start DFS from each cell; if word is found, return true.

return false;
    → If no path found after all cells, return false.

bool dfs(vector<vector<char>>& board, string& word, int r, int c, int i)
    → Private recursive helper function for DFS.

if (i == word.size()) return true;
    → Base case: matched entire word.

if (r < 0 || c < 0 || r >= ROWS || c >= COLS || board[r][c] != word[i])
    → If out of bounds or mismatch, return false.

char temp = board[r][c];
    → Save current cell value.

board[r][c] = '#';
    → Mark the cell as visited.

bool found = dfs(...) || dfs(...) || dfs(...) || dfs(...);
    → Explore all four directions.

board[r][c] = temp;
    → Restore cell’s original value for backtracking.

return found;
    → Return true if word found along this path.
------------------------------------------------------

⏱ Complexity:
- Time: O(m * n * 4^L), where m = rows, n = cols, L = word.length()
- Space: O(L), maximum depth of recursion stack.
------------------------------------------------------
*/

